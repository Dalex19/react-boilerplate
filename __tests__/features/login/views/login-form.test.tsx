import React from "react";
import LoginForm from "../../../../src/features/login/views/login-form";
import {fireEvent, screen, act, render, waitFor, cleanup} from "@testing-library/react";


const onLogin = jest.fn();

afterEach(() => {
    jest.clearAllMocks();
});

describe('<LoginForm />', function () {
    it('should render the <LoginForm /> component', function () {
        const { asFragment } = render(<LoginForm onLogin={onLogin} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should submit the form with username and password values', async function () {
        render(<LoginForm onLogin={onLogin} />)

        const username = screen.getByPlaceholderText('username');
        const password = screen.getByPlaceholderText('password');
        const submit = screen.getByText('Sign in');

        fireEvent.change(username, { target: { value: 'Laurence' }});
        fireEvent.change(password, { target: { value: '12345678' }});
        fireEvent.click(submit);

        await waitFor(() => expect(onLogin).toHaveBeenCalled());
        await waitFor(() => expect(onLogin).toHaveBeenCalledWith({ username: 'Laurence', password: '12345678' }));
    });

    it('should display required errors when value is not valid', async function () {
        render(<LoginForm onLogin={onLogin} />)
        const submit = screen.getByText('Sign in');

        fireEvent.click(submit);

        await waitFor(() => expect(screen.queryAllByRole('alert')).toHaveLength(2));
        expect(onLogin).not.toBeCalled();
    });

    it("should display min length error when password is invalid", async () => {
        render(<LoginForm onLogin={onLogin} />)

        const username = screen.getByPlaceholderText('username');
        const password = screen.getByPlaceholderText('password');

        fireEvent.change(username, { target: { value: 'Laurence' }});
        fireEvent.change(password, { target: { value: 'pass' }});

        fireEvent.submit(screen.getByText('Sign in'));

        expect(await screen.findAllByRole("alert")).toHaveLength(1);
        expect(onLogin).not.toBeCalled();

        expect(screen.getByRole('form')).toHaveFormValues({
            username: 'Laurence',
            password: 'pass',
        });
    });
});
