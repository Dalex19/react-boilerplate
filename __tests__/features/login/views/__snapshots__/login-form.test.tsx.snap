// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> should render the <LoginForm /> component 1`] = `
<DocumentFragment>
  <form
    class="card align-middle rounded-lg mt-52 container md:p-24 bg-white"
    name="login-form"
  >
    <div
      class="mx-auto"
    >
      <p
        class="pb-6 -mt-2 text-xl text-center"
      >
        Sign In
      </p>
    </div>
    <div
      class="login-form"
    >
      <input
        class="shadow-sm appearance-none focus:border-white border border-gray-100 rounded w-full py-2 mt-2 input"
        name="username"
        placeholder="username"
        type="text"
      />
      <input
        class="shadow-sm appearance-none focus:border-white border border-gray-100 rounded w-full py-2 mt-2 input"
        name="password"
        placeholder="password"
        type="password"
      />
      <input
        class="focus: shadow-sm w-full mt-6 btn bg-green-600 rounded text-white p-2"
        type="submit"
        value="Sign in"
      />
    </div>
  </form>
</DocumentFragment>
`;
